internal/repository/memory.go

```diff
@@ -0,0 +1,311 @@
+	// than len(p)
+	count := 0
+	j := 0 // index into p
+	for ; (j < len(p)) && (n.readCursor < len(data)); n.readCursor++ {
+		p[j] = data[n.readCursor]
+		count++
+		j++
+	}
+
+	// generate EOF if needed
+	var err error = nil
+	if n.readCursor >= len(data) {
+		err = io.EOF
+	}
+
+	return count, err
+}
+
+// Close implements io.Closer.Close
+func (n *nodeReaderWriter) Close() error {
+	n.readCursor = 0
+	n.writeCursor = 0
+	n.writing = false
+	return nil
+}
+
+// Write implements io.Writer.Write
+//
+// This implementation automatically creates the path n.path if it does not
+// exist. If it does exist, it is overwritten.
+func (n *nodeReaderWriter) Write(p []byte) (int, error) {
+
+	// guarantee that the path exists
+	_, ok := n.repo.Data[n.path]
+	if !ok {
+		n.repo.Data[n.path] = []byte{}
+	}
+
+	// overwrite the file if we haven't already started writing to it
+	if !n.writing {
+		n.repo.Data[n.path] = make([]byte, 0)
+		n.writing = true
+	}
+
+	// copy the data into the node buffer
+	count := 0
+	start := n.writeCursor
+	for ; n.writeCursor < start+len(p); n.writeCursor++ {
+		// extend the file if needed
+		if len(n.repo.Data) < n.writeCursor+len(p) {
+			n.repo.Data[n.path] = append(n.repo.Data[n.path], 0)
+		}
+		n.repo.Data[n.path][n.writeCursor] = p[n.writeCursor-start]
+		count++
+	}
+
+	return count, nil
+}
+
+// Name implements fyne.URI*Closer.URI
+func (n *nodeReaderWriter) URI() fyne.URI {
+
+	// discarding the error because this should never fail
+	u, _ := storage.ParseURI(n.repo.scheme + "://" + n.path)
+
+	return u
+}
+
+// NewInMemoryRepository creates a new InMemoryRepository instance. It must be
+// given the scheme it is registered for. The caller needs to call
+// repository.Register() on the result of this function.
+//
+// Since 2.0.0
+func NewInMemoryRepository(scheme string) *InMemoryRepository {
+	return &InMemoryRepository{
+		Data:   make(map[string][]byte),
+		scheme: scheme,
+	}
+}
+
+// Exists implements repository.Repository.Exists
+//
+// Since 2.0.0
+func (m *InMemoryRepository) Exists(u fyne.URI) (bool, error) {
+	if u.Path() == "" {
+		return false, fmt.Errorf("invalid path '%s'", u.Path())
+	}
+
+	_, ok := m.Data[u.Path()]
+	return ok, nil
+}
+
+// Reader implements repository.Repository.Reader
+//
+// Since 2.0.0
+func (m *InMemoryRepository) Reader(u fyne.URI) (fyne.URIReadCloser, error) {
+	if u.Path() == "" {
+		return nil, fmt.Errorf("invalid path '%s'", u.Path())
+	}
+
+	_, ok := m.Data[u.Path()]
+	if !ok {
+		return nil, fmt.Errorf("no such path '%s' in InMemoryRepository", u.Path())
+	}
+
+	return &nodeReaderWriter{path: u.Path(), repo: m}, nil
+}
+
+// CanRead implements repository.Repository.CanRead
+//
+// Since 2.0.0
+func (m *InMemoryRepository) CanRead(u fyne.URI) (bool, error) {
+	if u.Path() == "" {
+		return false, fmt.Errorf("invalid path '%s'", u.Path())
+	}
+
+	_, ok := m.Data[u.Path()]
+	if !ok {
+		return false, fmt.Errorf("no such path '%s' in InMemoryRepository", u.Path())
```

Comment:
I'm not sure this should return an error, the boolean answers the canread question with yes/no. An error should only be returned if the question cannot be answered

Reply:
Stating a file returns a struct and an error, whereas this is bool and an error. When a dev calls this, they will check the error and if not nil bubble it up, and if nil will check the boolean flag to decide an appropriate next step.

Reply:
Since this is an internal API, it ultimately comes down to "what is convenient for writing tests with". I found it to be convenient - for one thing, how else would you reliably generate CanRead errors in this context.

If someone feels strongly, I can change it, but I think it's fine as it stands.

Reply:
I am a little confused, Stuart's illustration to me shows that it is useful to return the error, so it can be bubbled up.
So yes, I agree, let's leave it.

Reply:
This behaviour is inconsistent with FileRepo, added commented there

Reply:
Is that necessarily a problem though? What constitutes an error will vary for different types of repositories.

Alternatively, perhaps we should explicitly document in the spec for `CanRead()` that non-existence is not an error?

If we need this property for testing purposes, we could always make it be a flag later.